asciidoctorj-core/src/main/java/org/asciidoctor/ast/AbstractNodeImpl.java

```diff
@@ -54,157 +63,211 @@ public DocumentRuby document() {
 
     @Override
     public DocumentRuby getDocument() {
-        return this.abstractNode.getDocument();
+        return (DocumentRuby) NodeConverter.createASTNode(getRubyProperty("document"));
+    }
+
+    public IRubyObject getRubyObject() {
+        return rubyNode;
     }
 
     @Override
     public String getNodeName() {
-        return this.abstractNode.getNodeName();
+        return getString("node_name");
     }
 
     @Override
     public boolean isInline() {
-        return this.abstractNode.isInline();
+        return getBoolean("inline?");
     }
 
     @Override
     public boolean isBlock() {
-        return this.abstractNode.isBlock();
+        return getBoolean("block?");
     }
 
     @Override
     public Map<String, Object> getAttributes() {
-        return this.abstractNode.getAttributes();
+        return new RubyAttributesMapDecorator((RubyHash) getRubyProperty("attributes"));
     }
 
     @Override
     public Object getAttr(Object name, Object defaultValue, boolean inherit) {
-        return this.abstractNode.getAttr(name, defaultValue, inherit);
+        return JavaEmbedUtils.rubyToJava(getRubyProperty("attr", name, defaultValue, inherit));
```

Comment:
I wonder if we should rename getAttr to getAttribute and isAttr to isAttribute. It's sort of in the Java style to have more formal names. What do you think would be more friendly?